Code/GraphMol/Depictor/EmbeddedFrag.cpp

```diff
@@ -341,11 +309,118 @@ int EmbeddedFrag::findNeighbor(
+  // setup
+  for (auto dai : d_attachPts) {
+    // find the neighbors that are already embedded for each of these atoms
+    RDKit::INT_VECT doneNbrs;
+    const auto &enbrs = d_eatoms[dai].neighs;
+    for (const auto nbrAtom :
+         dp_mol->atomNeighbors(dp_mol->getAtomWithIdx(dai))) {
+      if (std::find(enbrs.begin(), enbrs.end(),
+                    static_cast<int>(nbrAtom->getIdx())) == enbrs.end()) {
+        // we found a neighbor that is part of this embedded system
+        doneNbrs.push_back(nbrAtom->getIdx());
+      }
+    }
+    if (doneNbrs.empty()) {
+      d_eatoms[dai].normal = RDGeom::Point2D(1., 0.);
+      d_eatoms[dai].angle = -1.;
+    } else if (doneNbrs.size() == 1) {
+      auto nbid = doneNbrs.front();
+      d_eatoms[dai].nbr1 = nbid;
+      d_eatoms[dai].normal =
+          computeNormal(d_eatoms[dai].loc, d_eatoms[nbid].loc);
+    } else if (doneNbrs.size() == 2) {
+      auto nb1 = doneNbrs[0];
+      auto nb2 = doneNbrs[1];
+      d_eatoms[dai].nbr1 = nb1;
+      d_eatoms[dai].nbr2 = nb2;
+      d_eatoms[dai].angle =
+          computeAngle(d_eatoms[dai].loc, d_eatoms[nb1].loc, d_eatoms[nb2].loc);
+    } else if (doneNbrs.size() >= 3) {
+      // this is a pain - delegate it to a utility function
+      this->computeNbrsAndAng(dai, doneNbrs);
+    }
+  }
+}
+
+bool EmbeddedFrag::matchToTemplate(const RDKit::INT_VECT &ringSystemAtoms, unsigned int ring_count) {
+  CoordinateTemplates& coordinate_templates = CoordinateTemplates::getRingSystemTemplates();
+
+  // only look for an exact match to the ring system because our method of
+  // completing rings from a template isn't reliably better than not using
+  // a template at all
+  if (!coordinate_templates.hasTemplateOfSize(ringSystemAtoms.size())) {
+    return false;
+  }
+
+  // make a mol out of the induced subgraph using the ring system atoms
+  RDKit::RWMol rs_mol (*dp_mol);
+
+  // track original indices so that we can map template coordinates correctly
+  for (auto& at : rs_mol.atoms()) {
+    at->setProp(RDKit::common_properties::molAtomMapNumber, at->getIdx());
+  }
+
+  boost::dynamic_bitset<> rs_atoms (dp_mol->getNumAtoms());
+  for (auto aidx : ringSystemAtoms) {
+    rs_atoms.set(aidx);
+  }
+
+  rs_mol.beginBatchEdit();
+  for (auto& at : dp_mol->atoms()) {
+    if (!rs_atoms.test(at->getIdx())) {
+      rs_mol.removeAtom(at->getIdx());
+    }
+  }
+  rs_mol.commitBatchEdit();
+
+  // find template that this mol matches to, if any
+  RDKit::MatchVectType match;
+  std::shared_ptr<RDKit::ROMol> template_mol(nullptr);
+  for (const auto mol : coordinate_templates.getMatchingTemplates(ringSystemAtoms.size())) {
+    // To reduce how often we have to do substructure matches, check ring info and
+    // bond count first
+    if (mol->getNumBonds() != rs_mol.getNumBonds()) {
+      continue;
+    } else if (mol->getRingInfo()->numRings() != ring_count) {
+      continue;
+    }
+    RDKit::SubstructMatchParameters params;
```

Comment:
interesting, I wonder if there are other places where we really want "exact match" (graph mapping) rather than substructure search or MCS. I _presume_ that there may be some early exits possible in exact match.